enhancements/cluster-logging/cluster-logging-es-rollover-data-design.md

```diff
@@ -0,0 +1,253 @@
+---
+title: cluster-logging-es-rollover-data-design
+authors:
+  - "@jcantril"
+reviewers:
+  - "@bparees"
+  - "@ewolinetz"
+  - "@richm"
+  - "@portante"
+  - "@pavolloffay"
+  - "@sichvoge"
+approvers:
+  - "@ewolinetz"
+  - "@richm"
+creation-date: 2019-11-04
+last-updated: 2019-11-04
+status: implementable
+---
+
+# Cluster Logging Elasticsearch Rollover Data Design
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] ~~Graduation criteria for dev preview, tech preview, GA~~
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+This proposal alters the data design for storing logs in Elasticsearch to co-locate logs
+to fewer indices.  It additionally leverages the [Elasticsearch rollover api](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/indices-rollover-index.html) to
+help maintain the number of indices and shards in order to align with Elastic's 
+[performance](https://www.elastic.co/guide/en/elasticsearch/guide/2.x/scale.html) 
+and scaling [recommendations](https://www.elastic.co/blog/how-many-shards-should-i-have-in-my-elasticsearch-cluster).
+
+## Motivation
+
+The initial data design for Cluster Logging segments logs by OpenShift namespace in order to facilitate multi-tenant support and data curation.  This choice was made because
+index level security was the only feature available from an open source library.  It additionally facilitated curation as full indicies could be removed when the retention period
+expired.  This means, however, at any one time, there are at least "**$noOfNamespaces * $daysRetatined**" number 
+of indices maintained by the Elasticsearch server.  Each index can additionally be sharded to spread the load across the Elasticsearch nodes.  The end
+result of this shard explosion is the Elasticsearch cluster performance is not optimized and is not capable of efficiently processing logs or storing as many as it could. 
+
+Each index adds load and overhead (e.g. mapping, metadata) to the Elasticsearch cluster that needs to be tracked beyond the actual data.   Elasticsearch has recommendations for maximum shard size and the number of shards per node per allocated gigabyte of heap.  Cluster logging typically exceeds these recommendations for any OpenShift clusters that has significant log traffic.
+
+
+### Goals
+
+
+The goals of this proposals are:
+* Utilize a data design that aligns data schema with Elasticsearch's recommendations.
+* Expose data management policy as API in the `cluster-logging-operator` and `elasticsearch-operator` in support of Cluster Logging's mission to gather container and infrastructure logs
```

Comment:
is it a goal or non-goal to migrate existing customer log data to this new design?

Reply:
Seems like if this is implemented with the v7 rollout of Elasticsearch (okay, v6), then migrating from v5 to v6 could take care of moving old to new.

Reply:
it was a bit of a leading question but yes i expect data migration to be a goal.  but it should be stated clearly (or indicated that it is not)

Reply:
If we don't migrate from 5 to 7 now, then we'll have to have another release real soon that will migrate from 6 to 7 since 6 is going to be EOL "soon".

We have to solve the data migration no matter what, one way or another.  It would be great if this plan laid out how that migration is going to be handled.

Reply:
Added migration goal with additional details in the implementation section that follows